### PR TITLE
[SPARK-30766][SQL] Fix the timestamp truncation to the `HOUR` and `DAY` levels

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
In the PR, I propose to use Java 8 time API in timestamp truncations to the levels of `HOUR` and `DAY`. The problem is in the usage of `timeZone.getOffset(millis)` in days/hours truncations where the combined calendar (Julian + Gregorian) is used underneath.

### Why are the changes needed?
The change fix wrong truncations. For example, the following truncation to hours should print `0010-01-01 01:00:00` but it outputs wrong timestamp:
```scala
Seq("0010-01-01 01:02:03.123456").toDF()
    .select($"value".cast("timestamp").as("ts"))
    .select(date_trunc("HOUR", $"ts").cast("string"))
    .show(false)
+------------------------------------+
|CAST(date_trunc(HOUR, ts) AS STRING)|
+------------------------------------+
|0010-01-01 01:30:17                 |
+------------------------------------+
```

### Does this PR introduce any user-facing change?
Yes. After the changes, the result of the example above is:
```scala
+------------------------------------+
|CAST(date_trunc(HOUR, ts) AS STRING)|
+------------------------------------+
|0010-01-01 01:00:00                 |
+------------------------------------+
```

### How was this patch tested?
- Added new test to `DateFunctionsSuite`
- By `DateExpressionsSuite` and `DateTimeUtilsSuite`

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala [hunk 1]
```diff
@@ -713,32 +713,34 @@ object DateTimeUtils {
...
-    if (level == TRUNC_TO_MICROSECOND) return t
-    var millis = MICROSECONDS.toMillis(t)
-    val truncated = level match {
-      case TRUNC_TO_MILLISECOND => millis
-      case TRUNC_TO_SECOND =>
-        millis - millis % MILLIS_PER_SECOND
-      case TRUNC_TO_MINUTE =>
-        millis - millis % MILLIS_PER_MINUTE
-      case TRUNC_TO_HOUR =>
-        val offset = timeZone.getOffset(millis)
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: @MaxGekk, can you elaborate what's exactly wrong with `getOffset`?

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: It uses combined calendar - Julian + Gregorian. Look at the chain of calls:
TimeZone.getOffset -> TimeZone.getRawOffset ->ZoneInfo.getRawOffset -> ZoneInfo.getOffsets -> ZoneInfo.getOffsets(long date, int[] offsets, int type) -> SimpleTimeZone.getOffsets which uses combined calendar - for >= 1582 years Gregorian calendar, before 1582 - Julian calendar http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/687fd7c7986d/src/share/classes/java/util/SimpleTimeZone.java#l559:
```java
private static final Gregorian gcal = CalendarSystem.getGregorianCalendar();
...
BaseCalendar cal = date >= GregorianCalendar.DEFAULT_GREGORIAN_CUTOVER ?
                gcal : (BaseCalendar) CalendarSystem.forName("julian");
```

**Comment by HyukjinKwon** [comment]
Seems okay to me.

**Comment by MaxGekk** [comment]
@HyukjinKwon Anything else I can do here for merging the changes?

**Comment by MaxGekk** [comment]
@cloud-fan @HyukjinKwon @dongjoon-hyun Please, take a look at the PR.

File: sql/core/src/test/scala/org/apache/spark/sql/DateFunctionsSuite.scala [hunk 2]
```diff
@@ -856,4 +856,17 @@ class DateFunctionsSuite extends QueryTest with SharedSparkSession {
       TimeZone.setDefault(defaultTz)
     }
   }
+
+  test("date_trunc of old timestamps to hours and days") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/DateFunctionsSuite.scala]: let's add the JIRA ID

**Review by cloud-fan — APPROVED** [review]

**Review by HyukjinKwon — APPROVED** [review]

**Comment by MaxGekk** [comment]
jenkins, retest this, please

**Comment by cloud-fan** [comment]
thanks, merging to master/3.0!

**Pull request closed without merging**